src/util/memoizer.ts

```diff
@@ -0,0 +1,445 @@
+import * as ucum from '@lhncbc/ucum-lhc';
+import Immutable from 'immutable';
+import { Code, DateTime, Interval, Quantity, Ratio, Uncertainty } from '../datatypes/datatypes';
+import { decimalAdjust } from './math';
+import { convertUnit } from './units';
+
+const ucumUtilInstance = ucum.UcumLhcUtils.getInstance();
+
+/**
+ * Options for Memoize classes
+ */
+export interface IMemoizeOptions {
+  useCache?: boolean;
+  cacheMax?: number;
+}
+
+type Primitive = string | number | boolean | bigint | symbol | undefined | null;
+type KeyValuePair<KEY, VALUE> = { key: KEY; value: VALUE };
+
+/**
+ * Options for Memoize classes
+ */
+export interface IMemoizeOptions {
+  useCache?: boolean;
+  cacheMax?: number;
+}
+
+/**
+ * Let S, T, P, and Q be categories with the following diagram
+ *
+ * ```js
+ *  CACHE
+ *    ↑   ↘
+ *  P[][] → Q[]
+ *    ↑      ↓
+ *  S[][] → T[]
+ * ```
+ *
+ * This abstract class facilitates calculating S[][] → T[] via S[][] → P[][] → Q[] → T[] with the assumption that functors are chosen such that the diagram commutes.
+ *
+ * Moreover, this class optionally allows caching the calculation of P[][] → Q[] via the use of a unique CACHEKEY for P[][].
+ */
```

Comment:
Yup - you got it :)

Here's my thoughts on other descriptions...

One alternative description is that this is a React Higher-Order Component without React.  React documents Higher-Order Components at https://reactjs.org/docs/higher-order-components.html.

Here is the doDistinct implementation - I've added 3 boxes showing the analogy to a classic React component:

![image](https://user-images.githubusercontent.com/109760379/193936428-fce13e7c-53f9-4ad7-a322-69c4215eab26.png)

I've been hesitant to try to describe as a Higher-Order Component though.  React's documentation assumes a decent level of understanding of the library before explaining what a Higher-Order component is.

A second alternative description is that this is a lift in Haskell (see https://wiki.haskell.org/Lifting).  I don't think this goes as far as a Monadic lift.

In short, I don't think there is a universally accepted way to describe this technique, and I didn't want to assume a background in React, Haskell, etc.  The shortest interpretation I could think of was a category/functor diagram.

Let me give a concrete example.  Suppose I have a list of `Code`, and I want to perform doDistinct as in screenshot.  Then:

- `S[][]` is [Code[]]
- `P[][]` is [Immutable.Map[]]
- `Q[]` is Immutable.Map[]
- `T[]` is Code[]

And here is a description:

- `S[][]` -> `P[][]`: Lift the list of codes to a list of Immutable objects (doDistinct only takes 1 function param for second bracket)
- `P[][]` -> `Q[]`:  Perform a distinct operation on the Immutable objects to get a unique list of Immutable objects
- `Q[]` -> `T[]`: Drop the list of unique Immutable objects to a list of unique codes

I can see why the second bracket in `S[][]` and `P[][]` is confusing.  Consider the doExcept memoizer (red underline)

![image](https://user-images.githubusercontent.com/109760379/193939536-9fb4c701-a63f-471c-9870-fce339b06f2f.png)

The second set of brackets allows for describing functions and their lifts as taking an arbitrary number of parameters.  The doDistinct operation takes 1 list as a parameter.  The doExcept function takes 2 lists as parameters.

Reply:
I agree, referring to React HOC might confuse matters since it brings in a lot of irrelevant React details.  Haskell lifts might be helpful for people comfortable w/ Haskell. I don't know how many people that is.

So I think you're on the right track w/ your descriptions, but perhaps they just need a bit of what you've described above.  Could your examples above be generalized like this?

> For a single-argument function that takes in a list of `SomeType` (for example, the `doDistinct` function):
> * `S[][]` is `[SomeType[]]`
> * `P[][]` is `[Immutable.Map[]]`
> * `Q[]` is `Immutable.Map[]`
> * `T[]` is `SomeType[]`
>
> Given the above:
> * `S[][] -> P[][]` lifts the list of `SomeType` to a list of `Immutable.Map` objects
> * `P[][] -> Q[]` performs the passed in function (e.g., `doDistinct`) on the `Immutable.map` objects to get the result as a list of `Immutable.Map` objects (e.g., following on our example, the _distinct_ `Immutable.Map` objects)
> * `Q[] -> T[]` drops the resulting list of `Immutable.Map` objects back to a list of `SomeType`
>
> This allows for the primary algorithm (e.g., eliminating duplicates) to be performed more efficiently by operating on Immutable objects rather than the original objects.
>
> Note that the first dimension of the two-dimensional arrays `S[][]` and `P[][]` represents the argument list for the passed in function.  For example, in a function that takes two lists of`SomeType` as its arguments (e.g., `doExcept(SomeType[], SomeType[])`):
> * `S[][]` is `[SomeType[], SomeType[]]`
> * `P[][]` is `[Immutable.Map[], Immutable.Map[]]`
> * `Q[]` is `Immutable.Map[]`
> * `T[]` is `SomeType[]`

I know this is wordier, but I think it really helps to break it down for people (like me) who could benefit from being walked through it.